File: src/test/scala/workshop/part3/SuperComputeRouterTest.scala
package workshop.part3

import scala.language.postfixOps
import akka.actor._
import scala.concurrent.duration._
import workshop.work.{RiskyAdditionResult, RiskyAddition}
import workshop.AkkaSpec

class SuperComputeRouterTest extends AkkaSpec {

  val timeout: FiniteDuration = 1000 millis

  it should "compute 5 risky additions where none fail and where work is distributed on 5 routers" in  {
    suppressStackTraceNoise {
      val computeRouter = system.actorOf(Props(classOf[SuperComputeRouter]))

      val workList = Vector.fill(5) {
          RiskyAddition(1,1,150)
      }

      workList.foreach(computeRouter ! _)

      expectParallel(700) {
        var i = 0
        while(i < 5){
          expectMsgClass(timeout, classOf[RiskyAdditionResult])
          i = i + 1
        }
      }
    }
  }

}

File: src/test/java/workshop/part3/SuperComputeRouterTest.java
package workshop.part3;

import akka.actor.Actor;
import akka.actor.Props;
import akka.testkit.TestActorRef;
import org.junit.Test;
import workshop.AkkaTest;

import java.util.List;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;
import static workshop.work.Work.RiskyAddition;
import static workshop.work.Work.RiskyAdditionResult;
import static workshop.helpers.AkkaTestHelper.getGivenNumberOfResultsWithin;

public class SuperComputeRouterTest extends AkkaTest {

    @Test
    public void shouldCompute5RiskyAdditionsWhereNoneFailAndWhereWorkIsDistributedOn5Routers() {
        TestActorRef<Actor> computeRouter = TestActorRef.create(system, Props.create(SuperComputeRouter.class));

        List<RiskyAddition> workList = Stream.generate(() -> new RiskyAddition(1, 1, 150)).limit(5).collect(toList());

        workList.forEach(work -> computeRouter.tell(work, probe.ref()));

        List<RiskyAdditionResult> results = getGivenNumberOfResultsWithin(probe, workList.size(), 200, RiskyAdditionResult.class);
        results.forEach(result -> assertThat(result, is(new RiskyAdditionResult(2))));
    }
}
